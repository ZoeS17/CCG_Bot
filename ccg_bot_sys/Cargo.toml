[package]
authors = ["Zoe Kahala <>"]
name = "ccg_bot_sys"
license = "EUPL-1.2"
repository = "https://github.com/ZoeS17/CCG_Bot.git"
version = "0.1.0"
edition = "2021"
publish = false

[dependencies]
quote = "1"

[dependencies.syn]
version = "1"
features = ["derive", "full"]

[dev-dependencies]
trybuild = "1"

[lib]
name = "ccg_bot_sys"
path = "src/lib.rs"
proc-macro = true
