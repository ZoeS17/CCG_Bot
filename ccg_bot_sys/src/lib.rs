#![allow(dead_code)]

#[cfg(test)]
use trybuild;

use proc_macro::TokenStream;
use syn::{parse_macro_input, Item};

mod nonstable;

#[proc_macro_attribute]
pub fn nonstable(args: TokenStream, item: TokenStream) -> TokenStream {
    let args = parse_macro_input!(args as syn::AttributeArgs);
    let attr = nonstable::UnstableAttribute::from(args);

    match parse_macro_input!(item as Item) {
        Item::Type(item_type) => attr.expand(item_type),
        Item::Enum(item_enum) => attr.expand(item_enum),
        Item::Struct(item_struct) => attr.expand(item_struct),
        Item::Fn(item_fn) => attr.expand(item_fn),
        Item::Mod(item_mod) => attr.expand(item_mod),
        Item::Trait(item_trait) => attr.expand(item_trait),
        Item::Const(item_const) => attr.expand(item_const),
        Item::Static(item_static) => attr.expand(item_static),
        _ => panic!("unsupported item type"),
    }
}

#[test]
fn nonstable_without_feature() {
    let t = trybuild::TestCases::new();
    t.pass("src/tests/nonstable_without_feature.rs");
}

#[test]
fn nonstable_with_feature() {
    let t = trybuild::TestCases::new();
    t.pass("src/tests/nonstable_with_feature.rs");
}

#[test]
fn nonstable_without_issue() {
    let t = trybuild::TestCases::new();
    t.pass("src/tests/nonstable_without_issue.rs");
}

#[test]
fn nonstable_failure_with_use_item_type() {
    let t = trybuild::TestCases::new();
    t.compile_fail("src/tests/nonstable_fail_item_type.rs");
}

#[test]
fn nonstable_without_issue_or_feature() {
    let t = trybuild::TestCases::new();
    t.pass("src/tests/nonstable_without_issue_or_feature.rs");
}

#[test]
fn nonstable_with_feature_and_reason() {
    let t = trybuild::TestCases::new();
    t.pass("src/tests/nonstable_with_feature_and_reason.rs");
}

#[test]
fn nonstable_failure_with_non_str_feature() {
    let t = trybuild::TestCases::new();
    t.compile_fail("src/tests/nonstable_fail_non_str_feature.rs");
}

#[test]
fn nonstable_failure_with_non_str_issue() {
    let t = trybuild::TestCases::new();
    t.compile_fail("src/tests/nonstable_fail_non_str_issue.rs");
}
