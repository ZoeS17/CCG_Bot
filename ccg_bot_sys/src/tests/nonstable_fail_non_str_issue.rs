use ccg_bot_sys::nonstable;

#[nonstable(issue = 1)]
static NONSTABLE_STATIC_WITH_NON_STR_ISSUE: i32 = 0_i32;

pub fn main() {
    let _ = NONSTABLE_STATIC_WITH_NON_STR_ISSUE;
}