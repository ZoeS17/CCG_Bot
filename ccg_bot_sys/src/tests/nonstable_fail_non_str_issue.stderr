error: custom attribute panicked
 --> src/tests/nonstable_fail_non_str_issue.rs:3:1
  |
3 | #[nonstable(issue = 1)]
  | ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: explicit panic

error[E0425]: cannot find value `NONSTABLE_STATIC_WITH_NON_STR_ISSUE` in this scope
 --> src/tests/nonstable_fail_non_str_issue.rs:7:13
  |
7 |     let _ = NONSTABLE_STATIC_WITH_NON_STR_ISSUE;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
